opm/simulators/WellSwitchingLogger.cpp

```diff
@@ -0,0 +1,206 @@
+    {
+        int offset = displ[p];
+        int no_switches = 0;
+        MPI_Unpack(recv_buffer.data(), recv_buffer.size(), &offset,
+                   &no_switches, 1, MPI_INT, MPI_COMM_WORLD);
+
+        if ( no_switches == 0 )
+        {
+            continue;
+        }
+
+        std::vector<int> well_name_length(no_switches);
+
+        MPI_Unpack(recv_buffer.data(), recv_buffer.size(), &offset,
+                   well_name_length.data(), well_name_length.size(),
+                   MPI_INT, MPI_COMM_WORLD);
+
+        for ( int i = 0; i < no_switches; ++i )
+        {
+            char well_name[well_name_length[i]] = {};
+            MPI_Unpack(recv_buffer.data(), recv_buffer.size(), &offset,
+                       well_name, well_name_length[i], MPI_CHAR,
+                       MPI_COMM_WORLD);
+
+            std::array<char,2> fromto{{}};
+            MPI_Unpack(recv_buffer.data(), recv_buffer.size(), &offset,
+                       fromto.data(), 2, MPI_CHAR, MPI_COMM_WORLD);
+
+            logSwitch(well_name, fromto, p);
+        }
+    }
+}
+
+void WellSwitchingLogger::logSwitch(const char* name, std::array<char,2> fromto,
+                                    int rank)
+{
+            std::ostringstream ss;
+            ss << "Switching control mode for well " << name
+               << " from " << modestring[WellControlType(fromto[0])]
+               << " to " <<  modestring[WellControlType(fromto[1])]
+               << " on rank " << rank << std::endl;
+            OpmLog::info(ss.str());
+}
+#endif
+
+void WellSwitchingLogger::gatherDataAndLog()
+{
+
+#if HAVE_MPI
+    if(cc_.size() == 1)
+    {
+        return;
+    }
+
+    std::vector<int> message_sizes;
+    std::vector<int> well_name_length;
+    int message_size = calculateMessageSize(well_name_length);
+
+    if ( cc_.rank() == 0 ){
+        for(const auto& entry : switchMap_)
+        {
+            logSwitch(entry.first.c_str(), entry.second,0);
+        }
+
+        message_sizes.resize(cc_.size());
+    }
+
+    MPI_Gather(&message_size, 1, MPI_INT, message_sizes.data(),
+               1, MPI_INT, 0, MPI_COMM_WORLD);
+
+    std::vector<char> buffer(message_size);
+    packData(well_name_length, buffer);
+
+    std::vector<int> displ;
+
+    if ( cc_.rank() == 0){
+        // last entry will be total size of
+        displ.resize(cc_.size() + 1, 0);
+        std::partial_sum(message_sizes.begin(), message_sizes.end(),
+                         displ.begin()+1);
+    }
+    std::vector<char> recv_buffer;
+    if ( cc_.rank() == 0 ){
+        recv_buffer.resize(displ[cc_.size()]);
+    }
+
+    MPI_Gatherv(buffer.data(), buffer.size(), MPI_PACKED,
+                recv_buffer.data(), message_sizes.data(),
+                displ.data(), MPI_PACKED, 0, MPI_COMM_WORLD);
+    if ( cc_.rank() == 0 )
+    {
+        unpackDataAndLog(recv_buffer, displ);
+    }
+#endif
+}
+
+
+WellSwitchingLogger::~WellSwitchingLogger()
+{
+    gatherDataAndLog();
```

Comment:
I realize this is not the first time this pattern (do stuff in the destructor) is used in OPM, but I have become a bit wary of it, as it goes against one of the fundamental guidelines in C++: never throw in a destructor. I am not going to require this to change here, but we should think about it and see if there is another way to do this.